beacon_node/beacon_chain/src/sync_committee_verification.rs

```diff
@@ -0,0 +1,705 @@
+        self.signed_aggregate.message.contribution
+    }
+
+    /// Returns the underlying `signed_aggregate`.
+    pub fn aggregate(&self) -> &SignedContributionAndProof<T::EthSpec> {
+        &self.signed_aggregate
+    }
+}
+
+impl VerifiedSyncSignature {
+    /// Returns `Ok(Self)` if the `sync_signature` is valid to be (re)published on the gossip
+    /// network.
+    ///
+    /// `subnet_id` is the subnet from which we received this sync signature. This function will
+    /// verify that it was received on the correct subnet.
+    pub fn verify<T: BeaconChainTypes>(
+        sync_signature: SyncCommitteeSignature,
+        subnet_id: Option<SyncSubnetId>,
+        chain: &BeaconChain<T>,
+    ) -> Result<Self, Error> {
+        // Ensure sync committee signature is for the current slot (within a
+        // MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance).
+        //
+        // We do not queue future sync committee messages for later processing.
+        verify_propagation_slot_range(chain, &sync_signature)?;
+
+        // Sync signatures must be for a known block. If the block is unknown, we simply drop the
+        // sync committee message and do not delay consideration for later.
+        verify_head_block_is_known(chain, sync_signature.beacon_block_root)?;
+        let sync_subcommittee_size =
+            <<T as BeaconChainTypes>::EthSpec as EthSpec>::SyncCommitteeSize::to_usize()
+                .safe_div(SYNC_COMMITTEE_SUBNET_COUNT as usize)?;
+        let pubkey = chain
+            .validator_pubkey_bytes(sync_signature.validator_index as usize)?
+            .ok_or(Error::UnknownValidatorIndex(
+                sync_signature.validator_index as usize,
+            ))?;
+
+        let current_sync_committee = chain.head_current_sync_committee()?;
+        let mut subnet_positions = HashMap::new();
+        for (committee_index, validator_pubkey) in current_sync_committee.pubkeys.iter().enumerate()
+        {
+            if pubkey == *validator_pubkey {
+                let subcommittee_index = committee_index.safe_div(sync_subcommittee_size)?;
+                let position_in_subcommittee = committee_index.safe_rem(sync_subcommittee_size)?;
+                subnet_positions
+                    .entry(SyncSubnetId::new(subcommittee_index as u64))
+                    .or_insert_with(Vec::new)
+                    .push(position_in_subcommittee);
+            }
+        }
+
+        if let Some(subnet_id) = subnet_id {
+            if !subnet_positions.contains_key(&subnet_id) {
+                return Err(Error::InvalidSubnetId {
+                    received: subnet_id,
+                    expected: subnet_positions.keys().cloned().collect::<Vec<_>>(),
+                });
+            }
+        };
+
+        /*
+         * The sync committee message is the first valid message received for the participating validator
+         * for the slot, sync_signature.slot.
```

Comment:
The spec is unclear here, as Adrian Sutton pointed out this morning: https://github.com/ethereum/eth2.0-specs/issues/2471

As a first step I think we should modify the `observed_sync_contributors` to be keyed by `SlotSubcommitteeIndex`, like the signed contributions and proofs.

That issue looks like it will necessitate a spec change, but that can happen later

Reply:
Updated here: https://github.com/sigp/lighthouse/pull/2321/commits/bb7e9cf346d141f928fd0b6bc4a5e6ce3303dc29